### PR TITLE
Add Git credential helper exe

### PR DESCRIPTION
**Pull request by DazWorrall**
I used this gem to write a git credential helper for a project, it worked well and may be useful elsewhere, so this PR upstreams the code and makes it available via an exe bundled with the gem. I will follow up later with some details in the README of how to configure git to use it once I have done some further testing with a couple of different git setups.

### CODE REVIEW

File: github-authentication.gemspec [hunk 1]
```diff
@@ -36,7 +36,6 @@ Gem::Specification.new do |spec|
 
   spec.add_dependency("jwt", "~> 2.2")
 
-  spec.add_development_dependency("bundler", "~> 1.17")
```
Comment (DazWorrall) [inline github-authentication.gemspec:39]: This conflicts with #15, I think it's better to not be prescriptive about bundler in the gemspec.

**Comment by DazWorrall** [comment]
@doodzik are you[r team] still maintaining this gem, or was it handed off to someone else? Looking for a reviewer.

File: lib/github_authentication/git_credential_helper.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      puts("password=#{provider.token(seconds_ttl: min_cache_ttl)}")
+      puts('username=api')
+
+      0
+    end
+
+    private
+
+    def min_cache_ttl
+      # Tokens are valid for 60 minutes, allow a 10 minute buffer at the start of jobs
```
Comment (ChrisBr) [inline lib/github_authentication/git_credential_helper.rb:29]: [nit] It's not always job, is it?

```suggestion
      # Tokens are valid for 60 minutes, allow a 10 minute buffer
```

File: test/github_authentication/git_credential_helper_test.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    def stub_create_installation_access_token(token: SecureRandom.hex)
+      stub_request(:post, "https://api.github.com/app/installations/#{@installation_id}/access_tokens")
+        .to_return(
+          headers: { 'Content-Type' => 'application/json' },
+          body: { token: token, expires_at: 1.hour.from_now.iso8601 }.to_json,
+          status: 201,
+        )
+    end
+
+    def stub_parse_stdin(protocol: 'https', host: 'github.com', path: nil)
```
Comment (ChrisBr) [inline test/github_authentication/git_credential_helper_test.rb:54]: [nitpick] You could make STDIN an optional parameter and pass in a StringIO in tests. Then you don't have to stub a private method which couples your tests to the implementation a bit. And you would test your parsing method too.

```ruby
module GithubAuthentication
  class GitCredentialHelper
    def initialize(pem:, installation_id:, app_id:, storage: nil, stdin: STDIN)
      # ...
    end
  end
end

@helper = GitCredentialHelper.new(
  pem: @pem,
  app_id: @app_id,
  installation_id: @installation_id,
  stdin: StringIO.new("protocol=https\nhost=github.com"),
)
```

**Review by ChrisBr — APPROVED** [review]
Couple of nitpicks but looks sane.

File: lib/github_authentication/git_credential_helper.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    def handle_get
+      description = parse_stdin
+
+      unless description['protocol'] == 'https' && description['host'] == 'github.com'
+        warn("Unsupported description: #{description}")
+        return 2
+      end
+
+      puts("password=#{provider.token(seconds_ttl: min_cache_ttl)}")
```
Comment (doodzik) [inline lib/github_authentication/git_credential_helper.rb:21]: \nit readability
```suggestion
      token = provider.token(seconds_ttl: min_cache_ttl)
      puts("password=#{token}")
```

**Review by doodzik — APPROVED** [review]
Before merging, it would be good to add a little bit of a description to the readme on how to use the executable. It doesn't have to be too long. Other than that, LGTM.

**Comment by doodzik** [comment]
> @doodzik are you[r team] still maintaining this gem, or was it handed off to someone else? Looking for a reviewer.

@DazWorrall I don't think there has been any change in ownership for this repo. I'm happy to be pinged for reviews though.

**Pull request merged**